openedx_learning/apps/authoring/components/api.py

```diff
@@ -603,3 +607,61 @@ def _error_header(error: AssetError) -> dict[str, str]:
     )
 
     return HttpResponse(headers={**info_headers, **redirect_headers})
+
+
+def set_collections(
+    learning_package_id: int,
+    component: Component,
+    collection_qset: QuerySet[Collection],
+    created_by: int | None = None,
+) -> set[Collection]:
+    """
+    Set collections for a given component.
+
+    These Collections must belong to the same LearningPackage as the Component, or a ValidationError will be raised.
+
+    Modified date of all collections related to component is updated.
+
+    Returns the updated collections.
+    """
+    # Disallow adding entities outside the collection's learning package
+    invalid_collection = collection_qset.exclude(learning_package_id=learning_package_id).first()
+    if invalid_collection:
+        raise ValidationError(
+            f"Cannot add collection {invalid_collection.pk} in learning package  "
+            f"{invalid_collection.learning_package_id} to component {component} in "
+            f"learning package {learning_package_id}."
+        )
+    # Clear other collections for given component and add/keep only the ones in collection_qset
+    relations = CollectionPublishableEntity.objects.filter(
+        entity=component.publishable_entity
+    ).exclude(collection__in=collection_qset)
+    removed_collections = set()
+    if relations.exists():
+        relation = relations[0]
+        removed_collections = set(relation.collection for relation in relations)
+        # The normal `delete()` function deletes objects and triggers post delete signal for each object which is not
+        # required in this case as executing the signal handler once at the end should be more than enough.
+        # HACK: use private method `_raw_delete` deletes the objects without triggering the signals.
+        relations._raw_delete(relations.db)  # type: ignore[attr-defined] # pylint: disable=protected-access
+        # Trigger the signal manually for a single CollectionPublishableEntity object
+        post_delete.send(
+            sender=CollectionPublishableEntity,
+            instance=relation,
+            using=relations.db,
+            origin=relation,
+        )
```

Comment:
In edx-platform we have a `post_delete` [signal handler](https://github.com/open-craft/edx-platform/blob/eb7d2d1c23440cc9c035b6d1c0131ddd22c52a1e/openedx/core/djangoapps/content_libraries/signal_handlers.py#L165) for `CollectionPublishableEntity` which updates components collections in meilisearch index. In this specific case where we are operating on a single component and updating multiple collections we don't need to trigger reindexing on delete of each collection relation object as it slows down the whole operation a lot. 

I know that using private method to go around this issue is not the best solution but other option was to manage the collection indexing without `post_delete` signal of `CollectionPublishableEntity` model which I plan to investigate soon. Any advice will be appreciated.

Reply:
Instead of a signal handler for `receiver(post_delete, sender=CollectionPublishableEntity`, why don't you ignore the `post_delete` event (and emit several of them - no problem). Then, here in the code, instead of manually emitting one `post_delete` for `CollectionPublishableEntity` (which isn't really correct, because several were actually delete, and other parts of the system may want to know that), you can emit a custom signal for the entity itself, like `collection_changed`. Then the platform can listen to that single signal, and you don't have to suppress any "real" events or use private methods.

Reply:
> Instead of a signal handler for receiver(post_delete, sender=CollectionPublishableEntity, why don't you ignore the post_delete event

@bradenmacdonald I though the `post_delete` event is used to update component index (collections key) whenever a component is removed from a collection or a collection is deleted (all components are updated) but it seems like we also have `post_remove` action in `m2m_changed` signal doing the same thing. I'll investigate and test a bit more and update.

Reply:
Turns out that we can just remove `post_delete` handler for `CollectionPublishableEntity` as it is not required. The `post_remove` action in `m2m_changed` signal triggers single event for multiple removals in single `remove` call which is what we want.